common/achievement/get.ts

```diff
@@ -0,0 +1,67 @@
+import { prisma } from '../prisma';
+import { Achievement_template, User_achievement, achievement_action_type_enum, achievement_type_enum } from '../../graphql/generated';
+import { Step, Achievement, achievement_state } from '../../graphql/types/achievement';
+import { User } from '../user';
+
+const getUserAchievements = async (user: User): Promise<Achievement[]> => {
+    const userAchievements = await prisma.user_achievement.findMany({
+        where: { userId: user.userID },
+        include: { template: true },
+    });
+    const userAchievementGroups: { [group: string]: User_achievement[] } = {};
+    userAchievements.forEach((ua) => {
+        if (!userAchievementGroups[ua.template.group]) {
+            userAchievementGroups[ua.template.group] = [];
+        }
+        userAchievementGroups[ua.template.group].push(ua);
+    });
+    const achievements: Achievement[] = createFrontendAchievements(userAchievementGroups);
+    return achievements;
+};
+
+const createFrontendAchievements = (groups: { [group: string]: User_achievement[] }): Achievement[] => {
+    const achievements: Achievement[] = [];
+    for (const group in groups) {
+        const sortedGroupAchievements = groups[group].sort((a, b) => a.groupOrder - b.groupOrder);
+        const state: achievement_state =
+            groups[group].length === 0
+                ? achievement_state.INACTIVE
+                : groups[group].length === groups[group].length
+                ? achievement_state.COMPLETED
+                : achievement_state.ACTIVE;
```

Comment:
What is happening here? 😅 
Isn't `groups[group].length === groups[group].length` always true? Besides that, `groups[group].length === 0` can never happen right now.

Reply:
the first thing you are right, is completely rubbish. Don't know what got into me there...
for the length === 0 we have this since the frontend accepts inactive achievements and this is only possible if there is no user achievement initiated to a template. Not possible yes but Its part of the design, so the resolver allows it.

Reply:
the check for completed state checks if there is a achievement in the group that is not yet finished. else it is active state.